### PR TITLE
Level up

### PR DESCRIPTION
**Pull request by Cr3ativeZen**

### CODE REVIEW

File: apps/gameserver/src/Player.cpp [hunk 1]
```diff
@@ -829,16 +833,30 @@ void Player::Die()
...
+    if (amount == 0)
+        return;
+
+    if (GetLevel() >= 100)
     {
         spdlog::warn("ExpTable ends at lvl 100, exp will not be increased.");
         return;
     }
-    m_data.Exp += amount;
+    std::uint64_t exp_needed_for_next_level = g_nNeedExp[GetLevel()] * g_fNeedExpRatio[GetLevel()] / EXP_RATE;
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:846]: It would be better to do this calculation beforehand, at the server start and then re-use constant array here.

File: apps/gameserver/src/Player.cpp [hunk 2]
```diff
@@ -829,16 +833,30 @@ void Player::Die()
     WriteInSight(bango::network::packet(S2C_ACTION, "db", GetID(), AT_DIE));
 }
 
-void Player::UpdateExp(std::int64_t amount)
+void Player::UpdateExp(std::int64_t amount, bool send_update_to_client)
 {
-    if(GetLevel() >= 100) // ExpTable ends at lvl 100
+    if (amount == 0)
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:838]: What if amount is lower than 0?

File: apps/gameserver/src/Player.cpp [hunk 3]
```diff
@@ -829,16 +833,30 @@ void Player::Die()
...
+        return;
+
+    if (GetLevel() >= 100)
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:841]: We are currently not possible to decrease EXP if we are 100+ lv. Is that desired?

File: apps/gameserver/src/Player.cpp [hunk 4]
```diff
@@ -829,16 +833,30 @@ void Player::Die()
...
     {
         spdlog::warn("ExpTable ends at lvl 100, exp will not be increased.");
         return;
     }
-    m_data.Exp += amount;
+    std::uint64_t exp_needed_for_next_level = g_nNeedExp[GetLevel()] * g_fNeedExpRatio[GetLevel()] / EXP_RATE;
+    if (m_data.Exp + amount >= exp_needed_for_next_level)
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:847]: It should be possible to have 10 bars (100% of required exp) and still not level up. I think we should have `>` here?

File: apps/gameserver/src/Player.cpp [hunk 5]
```diff
@@ -829,16 +833,30 @@ void Player::Die()
...
+    {
+        std::int64_t leftover_exp = m_data.Exp + amount - exp_needed_for_next_level;
+        LevelUp();
+
+        if(leftover_exp > 0)
+            UpdateExp(leftover_exp, false);
+    }
+    else
+        m_data.Exp += amount;
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:856]: Correct me if I'm wrong but if we send update exp packet here then we do not need bool parameter in this function.

File: apps/gameserver/src/Player.cpp [hunk 6]
```diff
@@ -862,4 +880,33 @@ std::uint64_t Player::CalculateExp(std::uint64_t exp, std::uint8_t monster_level
...
+    m_data.SUPoint++;
+    m_data.PUPoint += GetAmountOfPUPointsOnLevelUp();
+    m_data.Exp = 0;
+    UpdatePropertyOnLevelUp();
+}
+
+std::uint8_t Player::GetAmountOfPUPointsOnLevelUp()
+{
+        return (
+            (GetLevel() >= 96 ? 12 :
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:897]: Where is this formula from? Are you sure it is not already defined in inix.h in a form of an array?

File: apps/gameserver/src/Player.cpp [hunk 7]
```diff
@@ -862,4 +880,33 @@ std::uint64_t Player::CalculateExp(std::uint64_t exp, std::uint8_t monster_level
...
+void Player::LevelUp()
+{
+    m_data.Level++;
+    m_data.SUPoint++;
+    m_data.PUPoint += GetAmountOfPUPointsOnLevelUp();
+    m_data.Exp = 0;
+    UpdatePropertyOnLevelUp();
+}
+
+std::uint8_t Player::GetAmountOfPUPointsOnLevelUp()
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:894]: Can be const

File: apps/gameserver/src/Player.cpp [hunk 8]
```diff
@@ -862,4 +880,33 @@ std::uint64_t Player::CalculateExp(std::uint64_t exp, std::uint8_t monster_level
...
+            (GetLevel() >= 72 ? 7 :
+                                5)))))));
+}
+
+void Player::UpdatePropertyOnLevelUp()
+{
+    SendProperty(P_LEVEL);
+    SendProperty(P_SUPOINT);
+    SendProperty(P_PUPOINT);
+    SaveAllProperty();
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:911]: I personally would prefer to have `SaveAllProperty` called outside of this function. We should avoid hiding database write/read in deeper parts of implementation.

File: apps/gameserver/include/Player.h [hunk 9]
```diff
@@ -158,7 +158,11 @@ class Player : public Character, public User
     void Tick() override;
     void Die() override;
 
-    void UpdateExp(std::int64_t amount);
+    void UpdateExp(std::int64_t amount, bool send_update_to_client = true);
     bool CanReciveExp();
     std::uint64_t CalculateExp(std::uint64_t exp, std::uint8_t monster_level);
+    void LevelUp();
+private:
+    std::uint8_t GetAmountOfPUPointsOnLevelUp();
```
Comment (lafreak) [inline apps/gameserver/include/Player.h:166]: This can be defined outside of Player class. I think it would not bloat player implementation. Please look up FIND_PU macro for example

File: apps/gameserver/include/Player.h [hunk 10]
```diff
@@ -158,7 +158,11 @@ class Player : public Character, public User
...
+    void UpdatePropertyOnLevelUp();
```
Comment (lafreak) [inline apps/gameserver/include/Player.h:167]: We are starting to get plenty of different Update methods. I think we will quickly end up in tens or hundreds of those in different flavors. Inix made it nicely with single UpdateProperty method which takes `kind` as parameter and `amount`. This way you can implement switch for all kind cases. This would not only increase/decrease the value but also take care of sending packet to the player. It would be nicely coupled for later use.

File: apps/gameserver/src/Player.cpp (hunk 6 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:897]: I found it in r3vo code. Seems to be right imo.

File: apps/gameserver/src/Player.cpp (hunk 5 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:856]: If we dont use Boolean here you will receive increase exp msg multiple times if u lvl up, which is not the case in Regular files.

File: apps/gameserver/src/Player.cpp (hunk 3 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:841]: There’s nothing to decrease if u can’t increase.

File: apps/gameserver/src/Player.cpp (hunk 2 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:838]: This Is the check if updateexp== 0 is sent. It will not do anything beside printing “ exp increased by 0”, if value is less than 0 then exp will decrease. Method name is “updateexp” not increase exp. We will use it later on to decrease exp for example after we click start from village.

File: apps/gameserver/src/Player.cpp (hunk 8 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:911]: It’s how it’s done in inix files.

File: apps/gameserver/src/Player.cpp (hunk 3 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp:841]: Entire gameserver is lacking unit tests and that is starting to bite us. There is something to decrease when you have 20 EXP left to lv100, kill monster and getp 500k exp. Then you are Lv. 100 with 499980 EXP. And you could decrease the exp in theory.

File: apps/gameserver/src/Player.cpp (hunk 5 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp:856]: Then maybe we should separate the message from actual calculation and increasement?

File: apps/gameserver/src/Player.cpp (hunk 6 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp:897]: To be verified (that source cant be trusted)

File: apps/gameserver/src/Player.cpp (hunk 3 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:841]: When u have 20 exp left to 100, then get 500k exp then u will get ur exp increased by 20 and rest of exp won't cuz line 850 call(UpdateExp) is after lvl up, so you are trying to update exp for lvl 100 player, which will pop the warning. The only problem is that you will get your exp increased clientside, and after relog it will be gone.

File: apps/gameserver/src/Player.cpp (hunk 4 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:847]: yep, done

File: apps/gameserver/src/Player.cpp (hunk 1 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:846]: done

File: apps/gameserver/src/Player.cpp (hunk 5 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Player.cpp:856]: done

File: apps/gameserver/src/Player.cpp [hunk 11]
```diff
@@ -862,4 +884,12 @@ std::uint64_t Player::CalculateExp(std::uint64_t exp, std::uint8_t monster_level
     //When monster_level > player_level, use g_nReviseExpA to calculate color ratio
     exp = (exp * (g_nReviseExpA[(GetLevel() - 1) / 10][level_difference] / 100.0) + exp);
     return exp;
-}
\ No newline at end of file
+}
+
+void Player::LevelUp()
+{
+    SendProperty(P_LEVEL, 1);
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:891]: What I meant with SendProperty was a little bit different. Look up how inix does it.

I would leave SendProperty to do one thing and do it good. **Only** send property to the client. Then no _amount_ parameter is needed to the function.

I would use this function in new, `UpdateProperty` method which takes _kind_ and _amount_. The UpdateProperty increase by amount would call SendProperty without amount parameter (player already has it increased).

File: apps/shared/inix/common.h [hunk 12]
```diff
@@ -29,7 +32,111 @@ static double g_fNeedExpRatio[] = {
 	5,5,5,5,5,5,5,5,5,5,5,5								//99-110
 };
 
-static std::uint8_t g_nReviseExpA[10][21] = {
+
+static constexpr std::uint64_t g_exp_table[MAX_LEVEL] = {
```
Comment (lafreak) [inline apps/shared/inix/common.h:36]: Cannot stay like that :weary:
I'd rather invest time writing proper compiletime calculation or use C style :( global variable with extern functionality.

Comment (lafreak) [reply apps/shared/inix/common.h:36]: https://stackoverflow.com/a/57214093/5480703

File: apps/shared/inix/common.h [hunk 13]
```diff
@@ -155,6 +263,23 @@ static int g_nAddDefLv[] = {
 	99, 99, 99, 99, 99, 99, 99, 99, 99, 99, 99
 };
 
+static std::uint8_t GET_PU_ON_LEVEL_UP(std::uint8_t level)
```
Comment (lafreak) [inline apps/shared/inix/common.h:266]: This function will be statically compiled into binary X times, where X is the number of cpp compile units (cpp files) which use this common.h header (I guess all use it).
I suggest to write macro, smilarly how FIND_PU_POINT does it.

File: apps/gameserver/src/Player.cpp [hunk 14]
```diff
@@ -829,16 +838,29 @@ void Player::Die()
...
 {
-    if(GetLevel() >= 100) // ExpTable ends at lvl 100
+    if (amount == 0)
+        return false;
+
+    if (GetLevel() >= MAX_LEVEL)
     {
         spdlog::warn("ExpTable ends at lvl 100, exp will not be increased.");
-        return;
+        return false;
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:849]: Log states that exp will not be increased, but this function is not only to increase but decrease as well

File: apps/gameserver/src/Player.cpp [hunk 15]
```diff
@@ -829,16 +838,29 @@ void Player::Die()
...
+    }
+
+    if (m_data.Exp + amount > GET_EXP_FOR_LEVEL(GetLevel()))
```
Comment (lafreak) [inline apps/gameserver/src/Player.cpp:852]: This if will not work and overflow in case of exp decrease. This shows that we got plenty of corner cases we do not cover "by hand". I think unit tests are required here.

Comment (lafreak) [reply apps/gameserver/src/Player.cpp:852]: In case we loose more exp than we have, this will give us 4kkkk exp = 100lv

File: apps/gameserver/src/Player.cpp (hunk 2 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp:843]: Yes, I meant what is happening if exp is _lower_ than 0, not equal. The behavior is not tested, as stated above

**Review by lafreak — APPROVED** [review]

File: apps/gameserver/src/Player.cpp (hunk 15 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp]: fixed

File: apps/shared/inix/common.h (hunk 12 above)
Comment (lafreak) [reply apps/shared/inix/common.h]: fixed

File: apps/gameserver/src/Player.cpp (hunk 11 above)
Comment (lafreak) [reply apps/gameserver/src/Player.cpp]: fixed

File: apps/gameserver/include/Player.h (hunk 10 above)
Comment (lafreak) [reply apps/gameserver/include/Player.h]: fixed

File: apps/gameserver/include/Player.h (hunk 9 above)
Comment (lafreak) [reply apps/gameserver/include/Player.h]: fixed
